src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2122,15 +2119,12 @@ internal override SemanticModel GetSemanticModelCore(SyntaxTree syntaxTree, bool
                 throw new ArgumentException(CSharpResources.SyntaxTreeNotFound, nameof(syntaxTree));
             }
 
-            if (SemanticModelProvider != null && useSemanticModelProviderIfNonNull)
-            {
-                Debug.Assert(!ignoreAccessibility);
-                return SemanticModelProvider.GetSemanticModel(syntaxTree, this);
-            }
-
-            return new SyntaxTreeSemanticModel(this, syntaxTree, ignoreAccessibility);
+            return SemanticModelProvider?.GetSemanticModel(syntaxTree, this, ignoreAccessibility) ?? CreateSemanticModel(syntaxTree, ignoreAccessibility);
```

Comment:
This code pattern implies that it's okay for `GetSemanticModel` to return `null` yet the declaration has it as a non-null return value. Think we should either change the definition to return `SemanticModel?` or change this code to check for `null` and throw / assert / etc ... if it's seen.

Reply:
Yeah, `SemanticModelProvider.GetSemanticModel` isn't annotated, so it shouldn't be returning null.

---
In reply to: [470112360](https://github.com/dotnet/roslyn/pull/46508#discussion_r470112360) [](ancestors = 470112360)